whisper/transcribe.py

```diff
@@ -8,15 +8,104 @@
 from .utils import exact_div, format_timestamp, optional_int, optional_float, str2bool, write_txt, write_vtt, write_srt
 
 if TYPE_CHECKING:
     from .model import Whisper
 
 
+def add_word_timestamps(
+    model: "Whisper",
+    tokenizer: Tokenizer,
+    mel: torch.Tensor,
+    num_frames: int,
+    segments: List[dict],
+    *,
+    medfilt_width: int = 7,
+    qk_scale: float = 1.0,
+):
+    if len(segments) == 0:
+        return
+
+    from dtw import dtw
+    from scipy.ndimage import median_filter
+
+    # install hooks on the cross attention layers to retrieve the attention weights
+    QKs = [None] * model.dims.n_text_layer
+    hooks = [
+        block.cross_attn.register_forward_hook(
+            lambda _, ins, outs, index=i: QKs.__setitem__(index, outs[-1])
+        )
+        for i, block in enumerate(model.decoder.blocks)
+    ]
+
+    tokens = torch.tensor(
+        [
+            *tokenizer.sot_sequence,
+            tokenizer.timestamp_begin,
+            *[t for segment in segments for t in segment["tokens"]],
+            tokenizer.timestamp_begin + mel.shape[-1] // 2,
+            tokenizer.eot,
+        ]
+    ).to(model.device)
+
+    with torch.no_grad():
+        model(mel.unsqueeze(0), tokens.unsqueeze(0))
+
+    for hook in hooks:
+        hook.remove()
+
+    weights = torch.cat(QKs)  # layers * heads * tokens * frames
+    weights = weights[:, :, :, : num_frames // 2].cpu()
+    weights = median_filter(weights, (1, 1, 1, medfilt_width))
+    weights = torch.tensor(weights * qk_scale).softmax(dim=-1)
+
+    w = weights / weights.norm(dim=-2, keepdim=True)
+    matrix = w.mean(axis=(0, 1))
+
+    alignment = dtw(-matrix.double().numpy())
+
+    jumps = np.pad(np.diff(alignment.index1s), (1, 0), constant_values=1).astype(bool)
+    jump_times = alignment.index2s[jumps] / TOKENS_PER_SECOND
+
+    if tokenizer.language in {"zh", "ja", "th", "lo", "my"}:
+        # These languages don't typically use spaces, so it is difficult to split words
+        # without morpheme analysis. Here, we instead split words at any
+        # position where the tokens are decoded as valid unicode points
+        split_tokens = tokenizer.split_tokens_on_unicode
+    else:
+        split_tokens = tokenizer.split_tokens_on_spaces
+
+    words, word_tokens = split_tokens(tokens[1:].tolist())
+
+    token_sources = np.repeat(np.arange(len(segments)), [len(s["tokens"]) for s in segments])
+    token_sources = [None] * len(tokenizer.sot_sequence) + list(token_sources)
+
+    time_offset = segments[0]["seek"] * HOP_LENGTH / SAMPLE_RATE
+    word_boundaries = np.pad(np.cumsum([len(t) for t in word_tokens]), (1, 0))
+    start_times = time_offset + jump_times[word_boundaries[:-1]]
+    end_times = time_offset + jump_times[word_boundaries[1:]]
+
+    for segment in segments:
+        segment["words"] = []
+
+    for i, (word, start, end) in enumerate(zip(words, start_times, end_times)):
+        if word.startswith("<|") or word.strip() in ".,!?、。":
```

Comment:
I think you're missing ？ and others. See:

https://github.com/jianfch/stable-ts/blob/main/stable_whisper/stabilization.py#L440%E2%88%92L443

Reply:
Would actually be convenient to actually re-insert the punctuation tokens so that concatenating all the words is the same as concatenating all the tokens. That would just make processing easier on the consumer end. For reference, Amazon Transcribe includes timestamped punctuation tokens in the results.

Reply:
I notice there is now a TODO comment leaning to the first approach:

```python
if word.startswith("<|") or word.strip() in ".,!?、。":  # TODO: expand
```

I'm not sure if you've already committed to that approach, but I would vote for not removing the punctuation, so that whether a consumer wants to traverse the entire content by token, by word, or by segment, they can and do it in either of these 3 ways and in all the content is there (the concatenation of each result is identical). Otherwise if I consume the results by word, I would need to simultaneously look up one of the other two results to cross reference the, and look for the bits that were omitted from the sequence. Here is how [Amazon Transcribe](https://docs.aws.amazon.com/transcribe/latest/dg/how-input.html) does it, for example.

On the other hand, if that is not persuasive, you might consider instead making it an option whether or not to strip out the punctuation.

(I note also that if you just left the punctuation in, the consumer would still have the ability to filter them out.)

Reply:
Thanks for the suggestions; I've updated so that the punctuation characters are attached to the adjacent words, while keeping the word's timestamps.

Reply:
Thanks, that looks good. I think if the `prepend_punctuations` and `append_punctuations` parameters were propagated in `transcribe()` and `cli()` that would be quite helpful, since then I could set them to empty strings to emulate the Amazon Transcribe behaviour.